File: test/fahasa_crawler.js
const puppeteer = require("puppeteer");
const fs = require("fs");

function removeDiacritics(str) {
  return str.normalize("NFD").replace(/[\u0300-\u036f]/g, "");
}

(async () => {
  const browser = await puppeteer.launch({
    headless: false, // Để kiểm tra giao diện thật (headless: "new" vẫn bị chặn ở 1 số site)
    args: [
      "--no-sandbox",
      "--disable-setuid-sandbox",
      "--disable-blink-features=AutomationControlled",
    ],
  });

  const page = await browser.newPage();

  await page.setUserAgent(
    "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 Chrome/138.0.0.0 Safari/537.36"
  );

  // Tránh bị phát hiện là bot
  await page.evaluateOnNewDocument(() => {
    Object.defineProperty(navigator, "webdriver", { get: () => false });
  });

  const totalPages = 1; // số trang muốn crawl
  const books = [];

  for (let p = 1; p <= totalPages; p++) {
    const url = `https://www.fahasa.com/sach-trong-nuoc.html?p=${p}`;
    console.log(`🔍 Đang xử lý trang ${p}: ${url}`);
    await page.goto(url, { waitUntil: "networkidle2" });
    await new Promise((resolve) => setTimeout(resolve, 2000));

    const productLinks = await page.$$eval("li div.item-inner", (items) =>
      items.map((el) => {
        const title =
          el.querySelector("h2.product-name-no-ellipsis a")?.innerText.trim() ||
          "";
        const url =
          el.querySelector("h2.product-name-no-ellipsis a")?.href || "";
        const img =
          el.querySelector("a.product-image img")?.getAttribute("data-src") ||
          "";
        const priceNew =
          el.querySelector("p.special-price span.price")?.innerText.trim() ||
          "";
        const priceOld =
          el.querySelector("p.old-price span.price")?.innerText.trim() || "";
        const discount =
          el.querySelector("span.discount-percent")?.innerText.trim() || "";
        const ratingCount =
          el.querySelector("div.rating-links")?.innerText.trim() || "0";

        return {
          title,
          url,
          image: img,
          priceNew,
          priceOld,
          discount,
          ratingCount,
        };
      })
    );

    for (const product of productLinks) {
      try {
        await page.goto(product.url, { waitUntil: "networkidle2" });
        await new Promise((resolve) => setTimeout(resolve, 1000));

        const description = await page
          .$eval("div.std", (el) => el.innerText.trim())
          .catch(() => "");

        const breadcrumb = await page.$$eval("ol.breadcrumb li", (items) =>
          items.map((i) => i.innerText.trim())
        );

        const category = breadcrumb.length >= 2 ? breadcrumb[1] : "khac";

        books.push({
          title: product.title,
          description,
          label: category,
        });

        console.log(`✅ ${product.title} [${category}]`);
        await new Promise((r) => setTimeout(r, 500));
      } catch (err) {
        console.warn(`⚠️ Lỗi khi crawl: ${product.url}`, err.message);
      }
    }
  }

  // Ghi file JSON
  fs.writeFileSync(
    "fahasa_books_puppeteer.json",
    JSON.stringify(books, null, 2),
    "utf-8"
  );

  console.log("\n📚 Một vài sách đầu tiên:");
  console.log(books.slice(0, 3)); // in 3 sách đầu tiên

  console.log(
    `🎉 Crawl xong ${books.length} sách. Lưu vào fahasa_books_puppeteer.json`
  );
  await browser.close();
})();

File: test/fahasa_crawler.py
import json
import time
from bs4 import BeautifulSoup
import undetected_chromedriver as uc


def remove_diacritics(text):
    return text.normalize("NFD").encode("ascii", "ignore").decode("utf-8")


def crawl_fahasa_books(total_pages=1, delay=1):
    books = []

    driver = uc.Chrome(
        headless=False,
        use_subprocess=True,
        options=None,
        args=[
            "--no-sandbox",
            "--disable-setuid-sandbox",
            "--disable-blink-features=AutomationControlled",
        ]
    )

    try:
        driver.set_page_load_timeout(30)

        # Tránh bị phát hiện là bot
        driver.execute_cdp_cmd(
            "Page.addScriptToEvaluateOnNewDocument",
            {
                "source": """Object.defineProperty(navigator, 'webdriver', {get: () => false})"""
            },
        )

        driver.execute_cdp_cmd(
            "Network.setUserAgentOverride",
            {
                "userAgent": "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 Chrome/138.0.0.0 Safari/537.36"
            },
        )

        for p in range(1, total_pages + 1):
            url = f"https://www.fahasa.com/sach-trong-nuoc.html?p={p}"
            print(f"🔍 Đang xử lý trang {p}: {url}")

            driver.get(url)
            time.sleep(delay)

            soup = BeautifulSoup(driver.page_source, "html.parser")
            items = soup.select("li div.item-inner")

            for item in items:
                try:
                    title_tag = item.select_one(
                        "h2.product-name-no-ellipsis a")
                    if not title_tag:
                        continue
                    title = title_tag.get_text(strip=True)
                    book_url = title_tag["href"]

                    # Crawl chi tiết
                    driver.get(book_url)
                    time.sleep(delay)

                    detail = BeautifulSoup(driver.page_source, "html.parser")
                    description_tag = detail.select_one("div.std")
                    description = description_tag.get_text(
                        strip=True) if description_tag else ""

                    breadcrumb = detail.select("ol.breadcrumb li a")
                    category = breadcrumb[1].get_text(
                        strip=True) if len(breadcrumb) >= 2 else "khác"

                    books.append({
                        "title": title,
                        "description": description,
                        "category": category
                    })

                    print(f"✅ {title} [{category}]")
                except Exception as e:
                    print(f"⚠️ Lỗi khi xử lý sách: {e}")

    finally:
        if driver:
            driver.quit()
            driver = None  # tránh __del__ gọi lại

    return books


if __name__ == "__main__":
    all_books = crawl_fahasa_books(total_pages=1, delay=2)

    with open("fahasa_books_cleaned.json", "w", encoding="utf-8") as f:
        json.dump(all_books, f, ensure_ascii=False, indent=2)

    print(f"\n📚 Đã lưu {len(all_books)} sách vào fahasa_books_cleaned.json")
    print("📌 Một vài sách đầu tiên:")
    print(json.dumps(all_books[:3], indent=2, ensure_ascii=False))
